File: November2021/Construct Binary Tree from Inorder and Postorder Traversal.cpp
class Solution {
    int idx;
    int solve1(vector<int> &inorder,int a,int start,int end)
    {
       for(int i=start;i<=end;i++)
           if(inorder[i]==a)
               return i;
        return -1;
    }
    TreeNode* solve(vector<int> &inorder, vector<int> &postorder, int start, int end)
    {
        if(start>end)
            return NULL;
     int val=postorder[idx];
     TreeNode *curr=new TreeNode(val);
     idx--;
     

        if(start==end)
            return curr;
     
     int pos=solve1(inorder,val,start,end);
     curr->right=solve(inorder,postorder,pos+1,end);
     curr->left=solve(inorder,postorder,start,pos-1);
        
     return curr;
    }
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        idx=inorder.size()-1;
        return(solve(inorder,postorder,0,inorder.size()-1));
    }
};

File: November2021/ConstructBinaryTreeFromInorderAndPostorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// TC : O(n)
// SC : O(n)
class Solution {
  private int pos =0;
  public TreeNode buildTree(int[] inorder, int[] postorder) {

    pos = postorder.length-1;

    Map<Integer, Integer> map = new HashMap<Integer,Integer>();
    for (int i=0;i<inorder.length;++i) {
      map.put(inorder[i], i);
    }

    TreeNode root = constructTree(inorder, postorder,  0, inorder.length-1, map );
    return root;
  }

  private TreeNode constructTree(int[] inorder, int[] postorder, int startIn, int endIn, Map<Integer, Integer> map){
    if(pos<0 || startIn> endIn){
      return null;
    }


    int val = postorder[pos];

    TreeNode node = new TreeNode(val);
    pos--;

    // indentify the position in the inorder tree
    int i= map.get(val);
    TreeNode right = constructTree(inorder, postorder, i+1, endIn, map);

    TreeNode left = constructTree(inorder, postorder, startIn, i-1, map);


    node.right = right;
    node.left = left;
    return node;
  }
}